Allow disabling OAuth2 consent prompt

maciek134 opened the pull request:
# Change summary

- Allow disabling the OAuth2 consent prompt

Fixes #3961

Checklist

- [x] This PR contains no AI generated code
- [x] book chapter included (if relevant) - section, not chapter
- [ ] design document included (if relevant)

Early PR to get some feedback. I'll add a few sentences in the book later.

maciek134 commented on server/lib/src/migration_data/dl13/schema.rs:
```diff
@@ -1128,6 +1136,7 @@ pub static ref SCHEMA_CLASS_OAUTH2_RS_DL9: SchemaClass = SchemaClass {
         Attribute::OAuth2RsOrigin,
         Attribute::OAuth2StrictRedirectUri,
         Attribute::OAuth2DeviceFlowEnable,
+        Attribute::OAuth2DisableConsentPrompt,
```
Also I'm not sure I was supposed to edit this class, it ofc worked on a dev setup with a clean db, but I noticed the `_DL9` suffix after making the PR

yaleman replied:
I'd definitely prefer `Attribute::OAuth2ConsentPromptDisable` as it uses the `thing -> sub-thing -> action` I'm encouraging

yaleman commented:
This'd be a good opportunity for you to document the schema migration modules and get @maciek134's feedback @Firstyear 😇

Firstyear commented on server/lib/src/idm/oauth2.rs:
```diff
@@ -874,6 +875,11 @@ impl Oauth2ResourceServersWriteTransaction<'_> {
+
+                let consent_prompt_enabled = ent
+                    .get_ava_single_bool(Attribute::OAuth2ConsentPromptEnable)
+                    .unwrap_or(true);
```
To answer your question re consent on public vs confidential clients: you can match on `type_` which tells you if the client is public. Then we allow this value to only go false on Basic clients.  

`let consent_prompt_enable == matches!(type_, OAuth2RSType::Public) || ent.get_ava .... `

maciek134 commented:
Ok, should be ready to review now. I've added the acp and moved the consent prompt bool to `OauthRSType::Basic`. Not sure if the integration test is in the right place, but everything seems to work.

I have to say this codebase is a pleasure to work with

maciek134 commented on server/lib/src/idm/oauth2.rs:
```diff
@@ -366,8 +366,9 @@ enum OauthRSType {
     Basic {
         authz_secret: String,
         enable_pkce: bool,
+        enable_consent_prompt: bool,
```
This became `enable_x` to be consistent with `pkce` - lmk if you prefer `consent_prompt_enable` to match the attribute

Firstyear replied:
Sounds good!

Firstyear commented:
> I have to say this codebase is a pleasure to work with

I'm really glad to hear that - I think there is a lot of "evolution" that has gone on that causes some rough edges around the place, but it's good to hear that you had a good experience 😅

Firstyear commented on server/lib/src/constants/uuids.rs:
```diff
@@ -510,6 +512,7 @@ pub const UUID_IDM_ACP_MESSAGE_MANAGE: Uuid = uuid!("00000000-0000-0000-0000-fff
 pub const UUID_IDM_ACP_MESSAGE_SENDER: Uuid = uuid!("00000000-0000-0000-0000-ffffff000078");
 pub const UUID_IDM_ACP_OAUTH2_CLIENT_ADMIN: Uuid = uuid!("00000000-0000-0000-0000-ffffff000079");
 pub const UUID_IDM_ACP_OAUTH2_ACCOUNT_ENROL: Uuid = uuid!("00000000-0000-0000-0000-ffffff000080");
+pub const UUID_IDM_ACP_OAUTH2_MANAGE_BASIC_V1: Uuid = uuid!("00000000-0000-0000-0000-ffffff000081");
```
Not needed anymore, it's some legacy/techdebt from old ideas that never came about, so you don't need the v1

Firstyear commented on server/lib/src/migration_data/dl13/access.rs:
```diff
@@ -724,6 +724,32 @@ lazy_static! {
 }
 
+lazy_static! {
+    pub static ref IDM_ACP_OAUTH2_MANAGE_BASIC: BuiltinAcp = BuiltinAcp {
```
Not needed, again that's a remnant of "how we used to do things", and it's no longer required.

Firstyear commented on server/lib/src/migration_data/dl13/access.rs:
```diff
@@ -724,6 +724,32 @@ lazy_static! {
+        classes: vec![
+            EntryClass::Object,
+            EntryClass::AccessControlProfile,
+            EntryClass::AccessControlCreate,
```
You don't need AccessControlCreate OR AccessControlDelete here.

Firstyear reviewed (commented):
Two minor changes regarding the access control, but looking really good.

Firstyear commented on server/lib/src/migration_data/dl13/access.rs:
```diff
@@ -724,6 +724,32 @@ lazy_static! {
+        search_attrs: vec![Attribute::OAuth2ConsentPromptEnable,],
+        modify_removed_attrs: vec![Attribute::OAuth2ConsentPromptEnable,],
+        modify_present_attrs: vec![Attribute::OAuth2ConsentPromptEnable,],
+        create_attrs: vec![Attribute::OAuth2ConsentPromptEnable,],
```
create_attrs define the attributes you can create while making an entry. There are a number of other subtelties too. 

Since we aren't setting the consent-prompt-enable during a new oauth2 client creation, we don't need this here. (Similar that's why you don't need AccessControlCreate)

the "AccessControlDelete" means that if you match the target you can delete the whole entry. Again, that's not the intent here, so it's not needed.

All that's needed is AccessControlSearch (so that you can see the consent prompt enable) and AccessControlModify (so you can assert a value is present on the entry, or assert the value is removed on the entry).

The present attrs defines what you can update as attributes. If you have present you can add a value, or assert a value already is present.

removed means you can remove a value, or assert it's already not present.

If you only have "present" the attribute becomes "append only". If you only have remove the attribute becomes "pop only". If you have both present and remove you can do an update in place etc.

Firstyear reviewed (approved):
Great work! Thank you so much!

maciek134 commented:
Soo I never asked and it's not in the book - who resolves discussions? Asking since you enabled auto-merge but it's complaining about unresolved comments and I default to letting reviewers resolve them.
